### PR TITLE
BUG: np array indexer modifed in iloc

### LINKED ISSUE #20852: np array indexer modified in iloc
#### Code Sample, a copy-pastable example if possible

```python
In [31]: import pandas as pd

In [32]: import numpy as np

In [33]: a = np.array([1, 2, -1])

In [34]: b = pd.DataFrame([100, 101, 102], index=[1, 2, 3], columns=['col'])

In [35]: b.iloc[a]
Out[35]:
   col
2  101
3  102
3  102

In [36]: a
Out[36]: array([1, 2, 2])


```
#### Problem description

The array a is modified after the iloc operation from array([ 1,  2, -1]) to array([1, 2, 2]).  The indexer should not be modified.

#### Expected Output
```python
In [39]: a
Out[39]: array([ 1,  2, -1])

```
#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.14.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-327.36.3.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: C
LANG: en_US.UTF-8
LOCALE: None.None

pandas: 0.22.0
pytest: 3.5.0
pip: 9.0.3
setuptools: 39.0.1
Cython: 0.28.2
numpy: 1.14.2
scipy: 1.0.1
pyarrow: 0.9.0
xarray: 0.10.2
IPython: 5.6.0
sphinx: 1.7.2
patsy: 0.5.0
dateutil: 2.7.2
pytz: 2018.4
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.2.2
openpyxl: 2.5.2
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.2.1
bs4: 4.3.2
html5lib: 0.999
sqlalchemy: 1.2.6
pymysql: None
psycopg2: 2.7.4 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by ydovzhenko**
- [ x ] closes #20852 
- [ x ] tests passed
- [ x ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: pandas/core/indexing.py [hunk 1]
```diff
@@ -2586,6 +2587,7 @@ def maybe_convert_indices(indices, n):
     IndexError : one of the converted indices either exceeded the number
         of elements (specified by `n`) OR was still negative.
     """
+    indices = deepcopy(indices)
```
Comment (jbrockmendel) [inline pandas/core/indexing.py]: Can this be done down on line 2601?  Avoid a copy if at all possible

**Comment by jbrockmendel** [comment]
Pls implement,met a test for this.

File: pandas/core/indexing.py (hunk 1 above)
Comment (ydovzhenko) [reply pandas/core/indexing.py]: Ok moving deepcopy to 2601 makes sense; will do. Can't see a way to avoid copying altogether.

Comment (WillAyd) [reply pandas/core/indexing.py]: Should be able to just use `.copy()` instead of importing deepcopy, no?

Comment (ydovzhenko) [reply pandas/core/indexing.py]: Yes, true. Will change to .copy()

File: pandas/core/indexing.py [hunk 2]
```diff
@@ -2596,6 +2598,8 @@ def maybe_convert_indices(indices, n):
 
     mask = indices < 0
     if mask.any():
+        # indices is np.array, so we don't use deepcopy
+        indices = np.copy(indices)
```
Comment (jreback) [inline pandas/core/indexing.py]: use ``indices.copy()``

File: pandas/core/indexing.py [hunk 3]
```diff
@@ -2596,6 +2598,8 @@ def maybe_convert_indices(indices, n):
 
     mask = indices < 0
     if mask.any():
+        # indices is np.array, so we don't use deepcopy
```
Comment (jreback) [inline pandas/core/indexing.py]: you don't need the comment

File: pandas/tests/indexing/test_iloc.py [hunk 4]
```diff
@@ -126,6 +126,24 @@ def test_iloc_getitem_neg_int(self):
                           typs=['labels', 'mixed', 'ts', 'floats', 'empty'],
                           fails=IndexError)
 
+    def test_iloc_array_not_mutating_negative_indices(self):
```
Comment (jreback) [inline pandas/tests/indexing/test_iloc.py]: can you add the issue number here as a comment

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a whatsnew for 0.24.0 bug fixes in indexing.

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/tests/indexing/test_iloc.py [hunk 5]
```diff
@@ -126,6 +126,24 @@ def test_iloc_getitem_neg_int(self):
...
+
+        # Test for issue 21867. Assert that iloc does not modify index array if
```
[the remark this answers is not in the document]
Comment (ydovzhenko) [reply pandas/tests/indexing/test_iloc.py]: Do you mean change the function name or the comment? Thanks!

File: pandas/core/indexing.py [hunk 6]
```diff
@@ -1,4 +1,5 @@
 # pylint: disable=W0223
+from copy import deepcopy
```
[the remark this answers is not in the document]
Comment (ydovzhenko) [reply pandas/core/indexing.py]: My bad.

File: pandas/tests/indexing/test_iloc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (WillAyd) [reply pandas/tests/indexing/test_iloc.py]: Comment

File: doc/source/whatsnew/v0.24.0.txt [hunk 7]
```diff
@@ -363,8 +363,7 @@ Indexing
 - ``DataFrame.__getitem__`` now accepts dictionaries and dictionary keys as list-likes of labels, consistently with ``Series.__getitem__`` (:issue:`21294`)
 - Fixed ``DataFrame[np.nan]`` when columns are non-unique (:issue:`21428`)
 - Bug when indexing :class:`DatetimeIndex` with nanosecond resolution dates and timezones (:issue:`11679`)
-
--
+- Bug when indexing with a numpy array containing negative values, fixed so the array is not mutated (:issue:`21867`)
```
Comment (WillAyd) [inline doc/source/whatsnew/v0.24.0.txt]: Can you be sure to capitalize `NumPy`? Can also reword to `Bug where indexing with a Numpy array containing negative values would mutate the indexer`

File: pandas/tests/indexing/test_iloc.py [hunk 8]
```diff
@@ -126,6 +126,23 @@ def test_iloc_getitem_neg_int(self):
                           typs=['labels', 'mixed', 'ts', 'floats', 'empty'],
                           fails=IndexError)
 
+    def test_iloc_array_not_mutating_negative_indices(self):
+
+        # GH 21867.
```
Comment (WillAyd) [inline pandas/tests/indexing/test_iloc.py]: Minor nit but don't need the period

**Review by WillAyd — CHANGES_REQUESTED** [review]
Very minor comments otherwise lgtm

File: pandas/tests/indexing/test_iloc.py [hunk 9]
```diff
@@ -126,6 +126,23 @@ def test_iloc_getitem_neg_int(self):
...
+            'A': [100, 101, 102],
+            'B': [103, 104, 105],
+            'C': [106, 107, 108]},
+            index=[1, 2, 3])
+        df.iloc[array_with_neg_numbers]
+        tm.assert_numpy_array_equal(array_with_neg_numbers, array_copy)
+
+        df.iloc[:, array_with_neg_numbers]
+        tm.assert_numpy_array_equal(array_with_neg_numbers, array_copy)
+
```
Comment (WillAyd) [inline pandas/tests/indexing/test_iloc.py]: Should only need one blank line here (think this will fail LINTing)

**Review by WillAyd — CHANGES_REQUESTED** [review]

**Review by WillAyd — APPROVED** [review]
Outside of the LINTing error this lgtm

File: pandas/tests/indexing/test_iloc.py (hunk 9 above)
Comment (ydovzhenko) [reply pandas/tests/indexing/test_iloc.py]: I can't find LINTing in the contributing guide. Can I LINT myself and how exactly do you guys do that?

Comment (TomAugspurger) [reply pandas/tests/indexing/test_iloc.py]: http://pandas-docs.github.io/pandas-docs-travis/contributing.html#python-pep8 has the quick version, but that doesn't check everything.

The actual command run during CI is `LINT=1 ci/lint.sh`, but that takes longer and may have additional dependencies.

https://travis-ci.org/pandas-dev/pandas/jobs/403382274#L2928 is the line causing this PR to fail.

Comment (TomAugspurger) [reply pandas/tests/indexing/test_iloc.py]: `flake8 pandas/tests/indexing/test_iloc.py` should work for you.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @ydovzhenko

**Pull request merged**